Add multiselect to Apps Forms

larkox opened the pull request:
#### Summary
Add multiselect support to Mobile

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-34194

#### Screenshots
![Screenshot from 2021-05-20 13-33-20](https://user-images.githubusercontent.com/1933730/119104558-b09f1b80-ba1c-11eb-8b73-83f3c076913c.png)
![Screenshot from 2021-05-20 13-33-28](https://user-images.githubusercontent.com/1933730/119104562-b137b200-ba1c-11eb-8790-d2a2e98ddeb9.png)
![Screenshot from 2021-05-20 13-33-59](https://user-images.githubusercontent.com/1933730/119104563-b137b200-ba1c-11eb-86e4-07fa1e22d6ea.png)
![Screenshot from 2021-05-20 13-40-33](https://user-images.githubusercontent.com/1933730/119104568-b1d04880-ba1c-11eb-9be3-a677a1ba0a7f.png)

#### Release Note
```release-note
Add multiselect support for Apps Forms
```

mattermod commented:
Building app in separate branch.

enahum commented:
@larkox please resolve conflicts

mickmister commented on app/components/autocomplete_selector/autocomplete_selector.tsx:
```diff
@@ -15,29 +14,42 @@ import {preventDoubleTap} from '@utils/tap';
+    },
+    getDynamicOptions?: (term: string) => Promise<ActionResult>,
+    label?: string,
+    placeholder?: string,
```
This prop changed from required to optional. This seems like it should be typed as optional but just pointing it out

larkox replied:
I don't remember 100% if I had any other motivation, but yes, this value should not be required.

mickmister commented on app/components/autocomplete_selector/autocomplete_selector.tsx:
```diff
@@ -48,26 +60,45 @@ export default class AutocompleteSelector extends PureComponent {
+        let selectedText = '';
+        const selected: DialogOption[] = [];
+
+        options.forEach((option) => {
+            if (selectedText !== '') {
+                selectedText += ', ';
+            }
+            selectedText += option.text;
+            selected.push(option);
+        });
```
Screenshot looks good. Nice job :+1:

mickmister commented on app/screens/selector_screen/selector_screen.tsx:
```diff
@@ -31,50 +30,79 @@ import {
+
+type DataType = DialogOption[] | Channel[] | UserProfile[];
+type Selection = DialogOption | Channel | UserProfile | DialogOption[] | Channel[] | UserProfile[];
+type Props = {
```
0/5 can we add a new line between these type declarations?

mickmister commented on app/screens/selector_screen/selector_screen.tsx:
```diff
@@ -31,50 +30,79 @@ import {
-        theme: PropTypes.object.isRequired,
-    };
+type State = {
+    data: DataType | {id: string; data: DataType}[],
+    loading: boolean;
+    searchResults: DialogOption[],
+    term: string;
+    multiselectSelected: DataType,
+}
```
nit: inconsistent field delimiters

mickmister commented on app/screens/selector_screen/selector_screen.tsx:
```diff
@@ -31,50 +30,79 @@ import {
+            data = props.data || [];
+        }
+
+        let multiselectSelected: DataType = [];
+        if (props.isMultiselect && props.selected && !([ViewTypes.DATA_SOURCE_USERS, ViewTypes.DATA_SOURCE_CHANNELS].includes(props.dataSource))) {
+            multiselectSelected = props.selected;
```
Just to be sure, this means you cannot initialize a multiselect list of users or channels right?

larkox replied:
Exactly. We should create a follow-up ticket to support it, but before we need to support this on webapp (user and channel multiselect is not yet supported).

mickmister reviewed (approved):
Awesome work @larkox :+1:

I have a few questions and some non-blocking formatting suggestions

larkox commented on app/components/autocomplete_selector/autocomplete_selector.tsx:
```diff
@@ -15,29 +14,42 @@ import {preventDoubleTap} from '@utils/tap';
+    onSelected?: ((item: DialogOption) => void) | ((item: DialogOption[]) => void),
+    helpText?: React.ReactNode,
+    errorText?: React.ReactNode,
+    roundedBorders?: boolean,
+    disabled?: boolean,
+    isMultiselect?: boolean;
+}
 
+type State = {
+    selectedText: string;
+    selected?: DialogOption | DialogOption[]
+}
```
Agree. For some reason I expected this to be handled by the linter.

larkox replied:
In the future it will be handled by the linter: https://github.com/mattermost/mattermost-mobile/pull/5514

larkox commented on app/screens/apps_form/apps_form_field.tsx:
```diff
@@ -24,18 +24,20 @@ export type Props = {
-    state = {
-        selected: null,
-    };
+    constructor(props:Props) {
+        super(props);
+
+        this.state = {};
+    }
```
Typescript does not like that. It feels like doing it that way "shadows" the actual state. In practice, probably does not, but still, typescript will consider `this.state` to have type `{}` instead of type `State`.

mickmister reviewed (approved):
LGTM :tada: Please re-request if needed after resolving conflicts or further changes

matthewbirtch commented:
@larkox is there a reason we omitted the checkbox from the rows here. If we're leveraging the multi-select pattern from new group messages, we should be consistent.

Here's a screenshot of what I mean:
<img src="https://user-images.githubusercontent.com/2040554/127347325-25ebc364-9ccf-49ce-abc2-aeaa473adfdd.PNG" width="414">

Also, for the button in the header that says `SELECT` - can we change that to `DONE`?

larkox commented:
@matthewbirtch Sure. Here is the new look.
![Screenshot from 2021-07-29 13-31-46](https://user-images.githubusercontent.com/1933730/127484984-bc480054-c763-41b9-8463-aa5bb93f42a4.png)
![Screenshot from 2021-07-29 13-32-04](https://user-images.githubusercontent.com/1933730/127484987-6b10be6f-cc43-4449-8cbd-532a35cf642c.png)
![Screenshot from 2021-07-29 13-32-20](https://user-images.githubusercontent.com/1933730/127484991-e293ecf1-6946-4469-b336-178685ada4fb.png)

larkox commented:
@mickmister Re-requesting your review since there are a few significant changes.

matthewbirtch commented:
> @matthewbirtch Sure. Here is the new look.

Thanks @larkox. I'll get a build going. Judging by the screenshots, it looks like the search inputs and the checkboxes are right up against the edges of the screen. I'll take a closer look at the build though

mattermod commented:
Building app in separate branch.

larkox commented:
@matthewbirtch They are indeed. I will look into a way to make them more similar to the create direct message menu.

matthewbirtch commented:
> @matthewbirtch They are indeed. I will look into a way to make them more similar to the create direct message menu.

Also, let me know what commands I should use to get this modal to display for review

larkox commented:
You need to have a environment with Apps enabled, and the test app running. Maybe @DHaussermann can help you with that part.
Once you have the environment set up, the command you have to run is `/test valid open_multiselect_form_modal`.

mickmister commented on app/screens/selector_screen/selector_screen.tsx:
```diff
@@ -41,6 +42,7 @@ import SelectedOptions from './selected_options';
 
 type DataType = DialogOption[] | Channel[] | UserProfile[];
 type Selection = DialogOption | Channel | UserProfile | DialogOption[] | Channel[] | UserProfile[];
+type MultiselectSelectedMap = Dictionary<DialogOption> | Dictionary<Channel> | Dictionary<UserProfile>
```
nit: missing semicolon at the end of this line

mickmister commented on app/screens/selector_screen/selector_screen.tsx:
```diff
@@ -131,35 +135,47 @@ export default class SelectorScreen extends PureComponent<Props, State> {
         switch (this.props.dataSource) {
         case ViewTypes.DATA_SOURCE_USERS: {
-            const currentList = this.state.multiselectSelected as UserProfile[];
+            const currentList = this.state.multiselectSelected as Dictionary<UserProfile>;
```
Should this be called `currentMap` now? Or something other than list, maybe `currentSelected`

mickmister reviewed (approved):
LGTM, nice UX improvements :+1:

matthewbirtch commented:
> You need to have a environment with Apps enabled, and the test app running. Maybe @DHaussermann can help you with that part.
> Once you have the environment set up, the command you have to run is `/test valid open_multiselect_form_modal`.

Ok, @DHaussermann let me know if this is something you can help with. We should look at creating an automated spinwick server with the apps environment somehow. Otherwise we can't test things from a UX perspective

DHaussermann commented:
@matthewbirtch in the interest of getting you something right now, I have the test app running here https://dkh4.test.mattermost.cloud and can DM you credentials.

@larkox we need something more permanent that this ☝️ I have started a discussion here https://community-release.mattermost.com/core/pl/bn8zpnmo1td9dcfodxdh53ddsh

matthewbirtch reviewed (changes requested):
@larkox I finally got around to reviewing this thanks to the test server from @DHaussermann. A few things I noticed:

**Initial screen:**
- The chevron for the inputs is _really_ tiny. I think we need to increase the size considerably (2x the size likely). I also wonder if it should point to the right (in iOS) since it pushes to a new screen that comes in from the right.
- We should fill out the rest of the screen with the grey background that is showing in the top half
![image](https://user-images.githubusercontent.com/2040554/128750500-77d5f1b9-1de4-4bef-9112-a388b0cb8536.png)

**Child screen**
- I would replace the `X` from the header and make it a back arrow (like a true child screen)
- Still seeing that we need to add the margin to the left of the selection circles here
![image](https://user-images.githubusercontent.com/2040554/128750846-6ad2e491-38ba-431c-94e1-eca9c5dab47b.png)

larkox commented:
@matthewbirtch Addressed your feedback.

- On Android the "select an option" string comes from the bottom, so I made a distinction between platform.
- Made the chevron bigger.
- The screens should now be filled in gray.
- The "select an option" screen has a back arrow instead of a cross.

Some things that I couldn't reproduce:
- The margin is there on my emulator.
- The close icon (now a back arrow) looks with the right color, not disabled.

Not sure if the things that I cannot reproduce has something to do with whether it is iOS or not. If you keep having those problems, let me know, and I will look deeper into that.

mattermod commented:
Building app in separate branch.

matthewbirtch reviewed (changes requested):
Thanks @larkox. A few more things from my test:

- Not sure if this is related to this PR, but the options in the autocomplete menu for slash commands have labels that aren't centered properly with the slash icon. They are sitting too high in the row.
<img src="https://user-images.githubusercontent.com/2040554/129094120-0fcacc73-1c6e-4ecd-bd88-c2fa71f9f1df.png" width="414">

- Should we treat the 'Test header' more like a header? I'm unsure what that is actually going to be, but it doesn't look like a heading for the page
- The chevrons look better for size, but maybe we can drop the opacity down a bit - something like 32% opacity? This should help calm it down
- Looks like the margin is there now, but it seems to be applied inconsistently. Not a big deal, but might as well be consistent. See the comparison of the 3 different screens below and notice the red line showing the difference in margin with the first static options

<img src="https://user-images.githubusercontent.com/2040554/129093114-806b70a3-d52d-4947-a6bd-4fce3d26b6e4.png" width="414">

- I see the back arrow now, but it is not displaying the iOS back arrow, it's displaying the android back arrow. Plus it's faded out?
<img src="https://user-images.githubusercontent.com/2040554/129093499-a2a1cc08-adaa-44a9-b8da-f0c097631f82.png" width="459">

larkox commented:
@matthewbirtch Everything should be "solved" with some caveats.

- Command texts: It is out of the scope of this PR, but I did a small improvement. Not sure if centered enough, but the baseline of the text is centered.
- The "test header" text is any introduction we want to set to the form. This is just a placeholder, but since it can be as big as we want (and even contain markdown) we should leave it as it is.
- Chevrons set to 32%
- Margins updated. It would be nice if you could check other "Select user" or "select channel" screens for unexpected results. I checked on Android, and it seems fine, but I don't know how it will look in iOS.
- I still cannot reproduce the arrow color problem, but I added another potential fix. Let me know if it is corrected.
- Now the app selects between an arrow to the left or a chevron to the left in case it is android or iOS for the back button.

mattermod commented:
Building app in separate branch.

matthewbirtch commented:
@larkox I'm not sure if the build finished. I don't see it posted in the QA: Builds channel. Did it fail?

larkox commented:
@matthewbirtch Might have. Rerunning just in case.

mattermod commented:
Building app in separate branch.

mattermod commented:
Building app in separate branch.

mattermod commented:
Building app in separate branch.

larkox commented:
@matthewbirtch The build was not getting generated because of a flaky test. Hopefully they get generated now.

@enahum For the flaky test (unrelated to this PR), I increased the wait time. Not sure if it will solve the problem in the long term. I haven't seen any other root cause.

matthewbirtch reviewed (changes requested):
@larkox I just tested the latest build and most of the issues are fixed. However, the back around icon used in the header is the wrong one. There is an icon the compass-icon font called `icon-arrow-back-ios`. Can you use that in the header instead of the `icon-chevront-left`?

<img src="https://user-images.githubusercontent.com/2040554/129388435-733851b7-a8cc-4c6e-9b79-19c46cd900e9.PNG" width="414">

Other than that, the other issues appear to be resolved.

mattermod commented:
Building app in separate branch.

larkox commented:
@matthewbirtch The back arrow now should be the correct one.

mickmister commented on app/actions/websocket/posts.test.js:
```diff
@@ -62,7 +62,7 @@ describe('Websocket Post Events', () => {
         PostSelectors.getPost.mockReturnValueOnce(false);
         mockServer.emit('message', message);
-        await TestHelper.wait(100);
+        await TestHelper.wait(300);
```
Is there something specific to this PR makes this change necessary?

larkox replied:
This is a naïve way to "fix" a flaky test that was giving us problems to generate the builds. It is not inherent to this PR, but if we don't have it, we might not be able to generate the builds on CI (since tests fail).

larkox commented:
@enahum Gentle reminder for your review on this.

enahum commented on app/components/autocomplete_selector/index.ts:
```diff
@@ -1,25 +1,30 @@
 import {setAutocompleteSelector} from 'app/actions/views/post';
 
 import AutocompleteSelector from './autocomplete_selector';
+import {Action, ActionResult, GenericAction} from '@mm-redux/types/actions';
+import {GlobalState} from '@mm-redux/types/store';
```
this are in fact types, you should import them as so

enahum commented on app/screens/selector_screen/selected_options/selected_options.tsx:
```diff
@@ -0,0 +1,85 @@
+    onRemove: (opt: DialogOption|UserProfile|Channel) => void;
+}
+
+export default function SelectedOptions(props: Props) {
```
One big issue seems that this component does not limit the amount of selected items and thus the view can grow limitless taking space from the list that sits below it on the screen. For me this is a blocker, look into adding some sort of ScrollView with a max height to avoid having this component filling up the screen.

larkox commented:
@enahum The changes are in place. Probably we should ask Matthew for another review when he is back. I made the maximum height 100 px, but it is as arbitrary as any other number.

On another UX issue, I feel the scroll view should scroll to bottom when new options are added or removed, but I did not implement it because I was not sure. What do you think? Should we defer to Matt?

enahum commented:
@larkox yes we can defer to @matthewbirtch on both things, but we could merge the PR and submit another one in case there are UX changes needed, what do you think?

matthewbirtch commented:
> @enahum The changes are in place. Probably we should ask Matthew for another review when he is back. I made the maximum height 100 px, but it is as arbitrary as any other number.

>On another UX issue, I feel the scroll view should scroll to bottom when new options are added or removed, but I did not implement it because I was not sure. What do you think? Should we defer to Matt?

Maximum height for what scroll view? Which piece are you referring to?

larkox commented:
@matthewbirtch We are referring to the piece where the selected items appear. When no item is selected, the scroll view of the list of all items fits all the available space. The moment you select one element, the list with all elements shrinks to make space for the selected element.
Before, the selected element list had no scroll, and will grow until it takes over the whole space, making the all element list unusable (if you were to add so many elements). Now, the selected element list has a maximum height, and became a scroll list. It will grow until 100px (completely arbitrary number, but felt right to me), and after that, will just scroll.

matthewbirtch commented:
> @matthewbirtch We are referring to the piece where the selected items appear. When no item is selected, the scroll view of the list of all items fits all the available space. The moment you select one element, the list with all elements shrinks to make space for the selected element.
> Before, the selected element list had no scroll, and will grow until it takes over the whole space, making the all element list unusable (if you were to add so many elements). Now, the selected element list has a maximum height, and became a scroll list. It will grow until 100px (completely arbitrary number, but felt right to me), and after that, will just scroll.

Got it. I'll take a look, but 100px sounds pretty decent to me.

mattermod commented:
Building app in separate branch.

matthewbirtch commented:
@larkox I agree that we should scroll the view to the bottom as you continue to add more selections so that you see it happening. As for the height of the scrollview, I think 100px is a reasonable max-height. 

One thing that would be good though, is to add a divider line between the chips scrollview area and the options below. Can we add a subtle divider here?

Also, I just noticed the following for the user and channel multi-select:
- Open the user mult-select view
- make selections
- hit `done`
- your selections show on the input
- tap the user input again to make changes
- it does not show the selections I previously made - it seems to clear them. 
 
This does not seem to happen for the static multi-select though.

larkox commented:
@matthewbirtch OK. I will try to add the scroll to bottom.

Regarding the divider, I will look how we do it in other parts of the code, to be consistent. Any idea of a divider used in other places that you want to have here?

Regarding the users being cleared, seems like a bug that I thought I had fixed. I will look into it.

Thank you very much!

matthewbirtch commented:
> Regarding the divider, I will look how we do it in other parts of the code, to be consistent. Any idea of a divider used in other places that you want to have here?

I think the same style divider that is used to separate each row in the list should work

larkox commented:
@matthewbirtch It does not reproduce for me (maybe another iOS / Android difference) but I applied what might be the fix. If you could give it another run to check, that would be awesome. If it doesn't get solved, I will look deeper into it.

mattermod commented:
Building app in separate branch.

mattermod commented:
Building app in separate branch.

matthewbirtch commented:
@larkox @DHaussermann is there another test server for me to use? The one I was using up to this point no longer works with the provided slash commands to launch the modal

matthewbirtch commented:
> @larkox @DHaussermann is there another test server for me to use? The one I was using up to this point no longer works with the provided slash commands to launch the modal

Nevermind.....ignore me :) I got it working

matthewbirtch reviewed (approved):
I think we're good to go now. Looks like it's resolved @larkox

mattermod commented:
Building app in separate branch.

mattermod commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/17977  
https://circleci.com/gh/mattermost/mattermost-mobile/17976

mattermod commented:
Artifact links:  
https://17977-70265724-gh.circle-artifacts.com/0/tmp/artifacts/Mattermost_Beta.apk  
https://17976-70265724-gh.circle-artifacts.com/0/tmp/artifacts/Mattermost_Beta.ipa

DHaussermann commented:
@larkox this work as I would expect. But I do see a couple issues.

1. `<--` arrow and `DONE` become unresponsive if the screen locks while the user picker, channel picker or Static option is open (**Android only** iOS modal seems unaffected.)
Steps:
- Run `/test valid open_multiselect_form_modal`
- Open the user picker or channel picker and select some options
- Turn off the screen (or set the phone down until the OS locks it for you which is more common)
- Unlock the phone to show the app again 
- Tap `<--` arrow and `DONE` to return to the main modal
**Observed**: 
A. Options are unresponsive
B. Sever logs and app console don't show any indication that the client is sending anything back 
C. User Must tap `<` twice to "exit" back to the modal

2. When re-opening the selection, the existing options are not preserved. the user must start again with building a list rather than just editing the existing selection. This issue does **not** occur for static options.  
Steps:
- Open the User or channel picker
- Select several of the options available
- Select `Done` and note that the selection has been saved
- Tap the user or channel option again to re-open the selection
**Observed**: The selection is now empty.     

Polish level issue maybe a P3 for later...
3. When using the searches, the first character entered will always show "No Results` for a brief moment then start to list them. A loading feedback here would have a nicer feel.

larkox commented:
For 1, I tried to reproduce in the emulator and I was not able. I will try on device.
For 2, I think there is already a ticket for this, which include reworking how we handle and codify the users and channels. It is expected, and will be out of the scope of this PR.
For 3, I think it can be a new ticket. Do you mind creating it?

DHaussermann commented:
Created https://mattermost.atlassian.net/browse/MM-38867 for the polish level issue 3.

larkox commented:
@DHaussermann I tried to reproduce 1 on device, and was not able.

@enahum Have you ever seen a similar behaviour?

enahum commented:
@larkox are you talking about the device becoming unresponsive? Yes I have seem that in the past, probably a lot of computation happening while the UI is being locked. @DHaussermann what device are you using for testing?

DHaussermann commented:
@enahum and @larkox testing is on a Galaxy S10 running Android 11 and the stock launcher (One UI v3.1)

enahum commented:
@larkox Not really as I have no idea what is going on behind the scenes, does this still reproduces? Is it only with specific steps? What about another screen being on top of the stack? What server is being used for testing?

jwilander commented:
@DHaussermann since @larkox is having difficulty reproducing and there's already a ticket for 3), are you ok merging this as is and we can follow up with any fixes afterwards?
